core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java

```diff
@@ -0,0 +1,100 @@
+import org.elasticsearch.action.index.TransportIndexAction;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.AutoCreateIndex;
+import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.transport.local.LocalTransport;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.test.cluster.TestClusterService;
+import org.junit.AfterClass;
+
+import java.util.HashSet;
+import java.util.concurrent.TimeUnit;
+
+public class DynamicMappingDisabledTests extends ESSingleNodeTestCase {
+
+    private static ThreadPool threadPool;
+    private TestClusterService clusterService;
+    private LocalTransport transport;
+    private TransportService transportService;
+    private IndicesService indicesService;
+    private ShardStateAction shardStateAction;
+    private ActionFilters actionFilters;
+    private IndexNameExpressionResolver indexNameExpressionResolver;
+    private AutoCreateIndex autoCreateIndex;
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        threadPool = new ThreadPool("DynamicMappingDisabledTests");
+        clusterService = new TestClusterService(threadPool);
+        transport = new LocalTransport(Settings.EMPTY, threadPool, Version.CURRENT, new NamedWriteableRegistry());
+        transportService = new TransportService(transport, threadPool);
+        indicesService = getInstanceFromNode(IndicesService.class);
+        shardStateAction = new ShardStateAction(Settings.EMPTY, clusterService, transportService, null, null);
+        actionFilters = new ActionFilters(new HashSet<>());
+        indexNameExpressionResolver = new IndexNameExpressionResolver(Settings.EMPTY);
+        autoCreateIndex = new AutoCreateIndex(Settings.EMPTY, indexNameExpressionResolver);
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        ThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+        threadPool = null;
+    }
+
+    public void testDynamicDisabled() {
+        Settings noDynamicIndexing = Settings.builder()
+            .put(MapperService.INDEX_MAPPER_DYNAMIC_SETTING, "false")
+            .build();
+
+        TransportIndexAction action = new TransportIndexAction(noDynamicIndexing, transportService, clusterService,
+            indicesService, threadPool, shardStateAction, null, null, actionFilters, indexNameExpressionResolver,
+            autoCreateIndex);
+
+        IndexRequest request = new IndexRequest("index", "type", "1");
+        request.source("foo", 3);
+
+        action.execute(request, new ActionListener<IndexResponse>() {
```

Comment:
If neither `onFailure` nor `onResponse` are mistakenly not executed, this test will still pass. I think that we should add something to avoid that possibility.